File: slimadapter/src/main/java/com/jhj/slimadapter/more/LoadMoreView.java
package com.jhj.slimadapter.more;

import android.support.annotation.IdRes;
import android.support.annotation.LayoutRes;
import android.view.View;

import com.jhj.slimadapter.holder.SlimViewHolder;

public abstract class LoadMoreView {

    public static final int STATUS_LOADING = 1;
    public static final int STATUS_FAIL = 2;
    public static final int STATUS_END = 3;

    private int mLoadMoreStatus = STATUS_LOADING;

    public void setLoadMoreStatus(int loadMoreStatus) {
        this.mLoadMoreStatus = loadMoreStatus;
    }

    public int getLoadMoreStatus() {
        return mLoadMoreStatus;
    }

    public void convert(SlimViewHolder holder) {
        switch (mLoadMoreStatus) {
            case STATUS_LOADING:
                visibleLoading(holder, true);
                visibleLoadFail(holder, false);
                visibleLoadEnd(holder, false);
                break;
            case STATUS_FAIL:
                visibleLoading(holder, false);
                visibleLoadFail(holder, true);
                visibleLoadEnd(holder, false);
                break;
            case STATUS_END:
                visibleLoading(holder, false);
                visibleLoadFail(holder, false);
                visibleLoadEnd(holder, true);
                break;
        }
    }

    private void visibleLoading(SlimViewHolder holder, boolean visible) {
        if (getLoadingViewId() != 0) {
            holder.getViewInjector().visibility(getLoadingViewId(), visible ? View.VISIBLE : View.GONE);
        }

    }

    private void visibleLoadFail(SlimViewHolder holder, boolean visible) {
        if (getLoadFailViewId() != 0) {
            holder.getViewInjector().visibility(getLoadFailViewId(), visible ? View.VISIBLE : View.GONE);
        }
    }

    private void visibleLoadEnd(SlimViewHolder holder, boolean visible) {
        if (getLoadEndViewId() != 0) {
            holder.getViewInjector().visibility(getLoadEndViewId(), visible ? View.VISIBLE : View.GONE);
        }
    }


    public void setLoadFailOnClickListener(SlimViewHolder holder, View.OnClickListener listener) {
        if (getLoadFailViewId() != 0) {
            holder.getViewInjector().clicked(getLoadFailViewId(), listener);
        }
    }

    /**
     * load more layout
     *
     * @return
     */
    public abstract @LayoutRes
    int getLayoutId();

    /**
     * loading view
     *
     * @return
     */
    protected abstract @IdRes
    int getLoadingViewId();

    /**
     * load fail view
     *
     * @return
     */
    protected abstract @IdRes
    int getLoadFailViewId();

    /**
     * load end view, you can return 0
     *
     * @return
     */
    protected abstract @IdRes
    int getLoadEndViewId();
}

File: slimadapterkt/src/main/java/com/zgdj/slimadapterkt/more/LoadMoreView.kt
package com.zgdj.slimadapterkt.more

import android.support.annotation.IdRes
import android.support.annotation.LayoutRes
import android.view.View

import com.zgdj.slimadapterkt.holder.SlimViewHolder

abstract class LoadMoreView {

    var loadMoreStatus = STATUS_LOADING
    private var listener: OnLoadMoreFailClickClListener? = null

    /**
     * load more layout
     *
     * @return
     */
    @get:LayoutRes
    abstract val layoutId: Int

    /**
     * loading view
     *
     * @return
     */
    @get:IdRes
    protected abstract val loadingViewId: Int

    /**
     * load fail view
     *
     * @return
     */
    @get:IdRes
    protected abstract val loadFailViewId: Int

    /**
     * load end view, you can return 0
     *
     * @return
     */
    @get:IdRes
    protected abstract val loadEndViewId: Int

    fun convert(holder: SlimViewHolder) {
        when (loadMoreStatus) {
            STATUS_LOADING -> {
                visibleLoading(holder, true)
                visibleLoadFail(holder, false)
                visibleLoadEnd(holder, false)
            }
            STATUS_FAIL -> {
                visibleLoading(holder, false)
                visibleLoadFail(holder, true)
                visibleLoadEnd(holder, false)
            }
            STATUS_END -> {
                visibleLoading(holder, false)
                visibleLoadFail(holder, false)
                visibleLoadEnd(holder, true)
            }
        }
    }

    private fun visibleLoading(holder: SlimViewHolder, visible: Boolean) {
        if (loadingViewId != 0) {
            holder.viewInjector.visibility(loadingViewId, if (visible) View.VISIBLE else View.GONE)
        }

    }

    private fun visibleLoadFail(holder: SlimViewHolder, visible: Boolean) {
        if (loadFailViewId != 0) {
            holder.viewInjector.visibility(loadFailViewId, if (visible) View.VISIBLE else View.GONE)
            holder.viewInjector.clicked(loadFailViewId, View.OnClickListener {
                loadMoreStatus = STATUS_LOADING
                convert(holder)
                listener!!.onClicked()
            })
        }
    }

    private fun visibleLoadEnd(holder: SlimViewHolder, visible: Boolean) {
        if (loadEndViewId != 0) {
            holder.viewInjector.visibility(loadEndViewId, if (visible) View.VISIBLE else View.GONE)
        }
    }


    fun setLoadFailOnClickListener(listener: OnLoadMoreFailClickClListener) {
        this.listener = listener
    }

    interface OnLoadMoreFailClickClListener {
        fun onClicked()
    }

    companion object {
        const val STATUS_LOADING = 1
        const val STATUS_FAIL = 2
        const val STATUS_END = 3
    }
}
